google/cloud/spanner/bytes.h

```diff
@@ -0,0 +1,157 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_SPANNER_GOOGLE_CLOUD_SPANNER_BYTES_H_
+#define GOOGLE_CLOUD_CPP_SPANNER_GOOGLE_CLOUD_SPANNER_BYTES_H_
+
+#include "google/cloud/spanner/version.h"
+#include "google/cloud/status_or.h"
+#include <array>
+#include <cstddef>
+#include <cstring>
+#include <iterator>
+#include <string>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+
+/**
+ * A representation of the Spanner BYTES type ... variable-length binary data.
+ *
+ * A `Bytes` value can be constructed from, and converted to (1) a sequence of
+ * octets, or (2) a base64-encoded US-ASCII `std::string`. The base64 alphabet
+ * is "A-Za-z+/", with "=" padding characters. Construction from a base64
+ * string can fail if the input string is unparsable.
+ *
+ * `Bytes` values can be compared for equality.
+ *
+ * All functions/operators should interpret the data as raw bytes (in contrast
+ * to the Spanner STRING type, where the data should be interpreted as Unicode
+ * characters).
+ */
+class Bytes {
+ public:
+  /// An empty sequence.
+  Bytes() {}
+
+  /// Construction from a sequence of octets.
+  ///@{
+  explicit Bytes(char const* p) : Bytes(p, p + std::strlen(p)) {}
+  template <typename InputIt>
+  Bytes(InputIt first, InputIt last) {
+    Encoder encoder(base64_rep_);
+    while (first != last) {
+      encoder.buf_[encoder.len_++] = *first++;
+      if (encoder.len_ == encoder.buf_.size()) encoder.Flush();
+    }
+    if (encoder.len_ != 0) encoder.FlushAndPad();
+  }
+  template <typename Container>
+  explicit Bytes(Container const& c) : Bytes(c.begin(), c.end()) {}
```

Comment:
I think we may want to use `std::begin(c)` and `std::end(c)` here, which will make this code more generic. Indeed, I think this change may then naturally handle the case of a string literal, because `std::begin` has the needed overload for arrays (https://en.cppreference.com/w/cpp/iterator/begin). So you can probably remove the `Bytes(char const*)` constructor??

Reply:
Done.  Thanks.